kinesis/dynamodb.py

```
@@ -0,0 +1,383 @@
+                        ":expected_ref": self.get_ref(),
+                    },
+                    ConditionExpression="#ref = :expected_ref",
+                    ReturnValues="ALL_NEW",
+                )
+
+                log.debug(f"{self.get_ref()} checkpointed on {shard_id}@{sequence}")
+                self._items[shard_id] = sequence
+
+            except ClientError as e:
+                if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                    raise Exception(f"{self.get_ref()} tried to checkpoint {shard_id} but does not own it")
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)